lighthouse-core/lib/csp-evaluator.js

```diff
@@ -0,0 +1,147 @@
+    'Please, use the Referrer-Policy header instead.',
+  deprecatedDisownOpener: 'disown-opener is deprecated since CSP3. ' +
+    'Please, use the Cross-Origin-Opener-Policy header instead.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @type {Record<number, Record<string, LH.IcuMessage>|string|LH.IcuMessage>} */
+const FINDING_TO_UI_STRING = {
+  [Type.MISSING_SEMICOLON]: UIStrings.missingSemicolon,
+  [Type.UNKNOWN_DIRECTIVE]: str_(UIStrings.unknownDirective),
+  [Type.INVALID_KEYWORD]: UIStrings.unknownKeyword,
+  [Type.MISSING_DIRECTIVES]: {
+    [Directive.BASE_URI]: str_(UIStrings.missingBaseUri),
+    [Directive.SCRIPT_SRC]: str_(UIStrings.missingScriptSrc),
+    [Directive.OBJECT_SRC]: str_(UIStrings.missingObjectSrc),
+  },
+  [Type.SCRIPT_UNSAFE_INLINE]: str_(UIStrings.unsafeInline),
+  [Type.NONCE_LENGTH]: str_(UIStrings.nonceLength),
+  [Type.DEPRECATED_DIRECTIVE]: {
+    [Directive.REFLECTED_XSS]: str_(UIStrings.deprecatedReflectedXSS),
+    [Directive.REFERRER]: str_(UIStrings.deprecatedReferrer),
+    [Directive.DISOWN_OPENER]: str_(UIStrings.deprecatedDisownOpener),
+  },
+  [Type.STRICT_DYNAMIC]: str_(UIStrings.strictDynamic),
+  [Type.UNSAFE_INLINE_FALLBACK]: str_(UIStrings.unsafeInlineFallback),
+  [Type.WHITELIST_FALLBACK]: str_(UIStrings.allowlistFallback),
+  [Type.REPORTING_DESTINATION_MISSING]: str_(UIStrings.reportingDestinationMissing),
+  [Type.REPORT_TO_ONLY]: str_(UIStrings.reportToOnly),
+};
+
+/**
+ * @param {Finding} finding
+ * @return {LH.IcuMessage|string}
+ */
+function getTranslatedDescription(finding) {
+  let result = FINDING_TO_UI_STRING[finding.type];
+  if (!result) {
+    log.warn('CSP Evaluator', `No translation found for description: ${finding.description}`);
+    return finding.description;
+  }
+
+  // Return if translated result found.
+  if (i18n.isIcuMessage(result)) return result;
+
+  // If result was not translated, that means `finding.value` is included in the UI string.
+  if (typeof result === 'string') return str_(result, {keyword: finding.value || ''});
+
+  // Result is a record object, UI string depends on the directive.
+  result = result[finding.directive];
+  if (!result) {
+    log.warn('CSP Evaluator', `No translation found for description: ${finding.description}`);
+    return finding.description;
+  }
+
+  return result;
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Finding>}
+ */
+function evaluateRawCspForFailures(rawCsps) {
+  return evaluateForFailure(rawCsps.map(c => new Parser(c).csp));
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Finding>}
+ */
+function evaluateRawCspForWarnings(rawCsps) {
+  return evaluateForWarnings(rawCsps.map(c => new Parser(c).csp));
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Array<Finding>>}
+ */
+function evaluateRawCspForSyntax(rawCsps) {
+  return evaluateForSyntaxErrors(rawCsps.map(c => new Parser(c).csp));
```

Comment:
Is there any significant overhead with `Parser`? Should we explore restructuring the API here so that we can provide `rawCsp` just once, and in return get details separated into warnings/failures/syntax all at once?

Reply:
Old code but same logic:
https://github.com/google/csp-evaluator/blob/f40c18cbe35f3ac5a6c7d04da64f7078b9fca4b1/evaluator/parser.js#L51-L99

It isn't a huge overhead, but I don't see any reason not to go with your suggestion other than restructuring the code a bit.

Reply:
I think I'm going to leave these because it made testing easier.